File: Frontend/StudyVera.FrontEnd/StudyVera.FrontEnd/Services/UserWeeklyGoalService.cs
﻿using Blazored.LocalStorage;
using StudyVera.FrontEnd.Models.Common;
using StudyVera.FrontEnd.Models.UserActivityHistories;
using StudyVera.FrontEnd.Models.UserWeeklyGoals;
using StudyVera.FrontEnd.Services.Concrats;
using StudyVera.FrontEnd.Services.Helpers;
using StudyVera.FrontEnd.Utilities;
using System.Text.Json;

namespace StudyVera.FrontEnd.Services;

public class UserWeeklyGoalService : ServiceHelper, IUserWeeklyGoalService
{
    private readonly HttpClient _client;
    private readonly ILocalStorageService _localStorage;
    private string _baseUrl = $"{AppConsts.ApiBaseUrl}weekly-goal";
    public UserWeeklyGoalService(ILocalStorageService localStorage, HttpClient client) : base(localStorage, client)
    {
        _client = client;
        _localStorage = localStorage;
    }

    public async Task<UserWeeklyGoalDto> GetUserWeeklyGoal()
    {
        try
        {
            await AddAuthorizationHeader();

            var response = await _client.GetAsync($"{_baseUrl}/summary");

            if (!response.IsSuccessStatusCode)
            {
                await response.HandleError();
            }

            var content = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<UserWeeklyGoalDto>(content,
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true })
                ?? new UserWeeklyGoalDto();
        }
        catch (HttpRequestException)
        {
            throw new Exception("İnternet bağlantınızı kontrol edin, sunucuya ulaşılamıyor.");
        }
        catch (Exception ex)
        {
            throw;
        }
    }
}

File: Frontend/StudyVera/src/app/services/user-weekly-goal.service.ts
// src/app/services/user-weekly-goal.service.ts

import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { UserWeeklyGoalDto } from '../models/user-weekly-goal.model';

@Injectable({
  providedIn: 'root'
})
export class UserWeeklyGoalService {
  private http = inject(HttpClient);
  
  private baseUrl = `${environment.apiUrl}/weekly-goal`;

  getUserWeeklyGoal(): Observable<UserWeeklyGoalDto> {
    return this.http.get<UserWeeklyGoalDto>(`${this.baseUrl}/summary`).pipe(
      catchError((error: HttpErrorResponse) => {
        if (error.status === 0) {
          return throwError(() => new Error('İnternet bağlantınızı kontrol edin, sunucuya ulaşılamıyor.'));
        }
        
        return throwError(() => error);
      })
    );
  }
}